flake8_trio.py

```diff
@@ -255,6 +255,135 @@ def check_for_trio100(self, node: Union[ast.With, ast.AsyncWith]) -> None:
                 )
 
 
+# Never have an except Cancelled or except BaseException block with a code path that
+# doesn't re-raise the error
+class Visitor103_104(ast.NodeVisitor):
+    def __init__(self) -> None:
+        super().__init__()
+        self.problems: List[Error] = []
+
+        self.except_name: Optional[str] = None
+        self.unraised: bool = False
+
+    def visit_ExceptHandler(self, node: ast.ExceptHandler):
+        def has_exception(node: Optional[ast.expr]):
+            return (isinstance(node, ast.Name) and node.id == "BaseException") or (
+                isinstance(node, ast.Attribute)
+                and isinstance(node.value, ast.Name)
+                and node.value.id == "trio"
+                and node.attr == "Cancelled"
+            )
+
+        outer_unraised = self.unraised
+        exc_node = None
+
+        if isinstance(node.type, ast.Tuple):
+            for element in node.type.elts:
+                if has_exception(element):
+                    exc_node = element
+                    break
+        elif has_exception(node.type):
+            exc_node = node.type
+
+        if exc_node is not None:
+            self.except_name = node.name
+            self.unraised = True
+
+        self.generic_visit(node)
+
+        if exc_node is not None:
+            if self.unraised:
+                self.problems.append(
+                    make_error(TRIO103, exc_node.lineno, exc_node.col_offset)
+                )
+
+        self.unraised = outer_unraised
+
+    def visit_Raise(self, node: ast.Raise):
+        # if there's an exception that must be raised
+        # and none of the valid ways of re-raising it is done
+        if self.unraised and not (
+            # bare except
+            node.exc is None
+            # re-raised by name
+            or (isinstance(node.exc, ast.Name) and node.exc.id == self.except_name)
+            # new valid exception raised
+            or (
+                isinstance(node.exc, ast.Call)
+                and (
+                    (
+                        isinstance(node.exc.func, ast.Name)
+                        and node.exc.func.id == "BaseException"
+                    )
+                    or (
+                        isinstance(node.exc.func, ast.Attribute)
+                        and isinstance(node.exc.func.value, ast.Name)
+                        and node.exc.func.value.id == "trio"
+                        and node.exc.func.attr == "Cancelled"
+                    )
+                )
+            )
```

Comment:
Pretty sure that it's never valid to raise a new exception!

For example this would allow replacing a `Cancelled` with a `BaseException`, which would definitely break things.

Reply:
ooh right, yeah makes sense that would screw with trio - as would the reverse.
Though
```python
except trio.Cancelled:
  raise trio.Cancelled()
```
is fine in theory right? But if there's never/rarely any good reason to do that (instead of modifying the exception) can just simplify the logic.

Reply:
[Nope: per the Trio docs](https://trio.readthedocs.io/en/stable/reference-core.html#a-simple-timeout-example)

> every [Cancelled](https://trio.readthedocs.io/en/stable/reference-core.html#trio.Cancelled) object carries an invisible marker that makes sure that the cancel scope that triggered it is the only one that will catch it.

So you really *can't* replace the exception without breaking things.

Reply:
Lol, gotta love invisible markers.

Fixed :+1: